File: python/examples/pro-tier/create-comparison-table/create_comparison_table.py
import argparse
import sys
from collections import defaultdict
from pathlib import Path
from typing import Any

import cs2cap
from cs2cap.rest import ApiException

EXAMPLES_ROOT = Path(__file__).resolve().parents[2]
if str(EXAMPLES_ROOT) not in sys.path:
    sys.path.insert(0, str(EXAMPLES_ROOT))

from _shared.api import parse_ndjson_lines  # noqa: E402
from _shared.auth import build_configuration, load_api_key, require_real_api_key  # noqa: E402
from _shared.render import format_money, format_percent, render_table  # noqa: E402


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Stream a bounded prices snapshot and enrich selected items with batch bids.",
    )
    parser.add_argument("--max-rows", type=int, default=500, help="Maximum NDJSON rows to parse.")
    parser.add_argument("--top", type=int, default=8, help="Number of spread rows to print.")
    return parser


def as_int(value: object) -> int | None:
    return value if isinstance(value, int) else None


def as_str(value: object) -> str | None:
    return value if isinstance(value, str) else None


def main() -> int:
    args = build_parser().parse_args()

    try:
        bearer_token = load_api_key("CS2C_API_KEY", start=Path(__file__))
        require_real_api_key(bearer_token)
    except RuntimeError as exc:
        print(str(exc), file=sys.stderr)
        return 1

    configuration = build_configuration(bearer_token)

    try:
        with cs2cap.ApiClient(configuration) as client:
            prices_api = cs2cap.PricesApi(client)
            bids_api = cs2cap.BidsApi(client)

            snapshot = prices_api.stream_full_prices_snapshot()
            price_rows = parse_ndjson_lines(snapshot, max_rows=max(args.max_rows, 1))

            grouped: dict[int, list[dict[str, Any]]] = defaultdict(list)
            names: dict[int, str] = {}
            for row in price_rows:
                item_id = as_int(row.get("item_id"))
                name = as_str(row.get("market_hash_name"))
                if item_id is None or name is None:
                    continue
                grouped[item_id].append(row)
                names[item_id] = name

            item_ids = list(grouped)[: max(args.top * 2, args.top)]
            bid_response = bids_api.batch_bid_lookup(
                cs2cap.BatchBidsRequest(item_ids=item_ids, currency="USD")
            )

    except ApiException as exc:
        print(f"API request failed: {exc}", file=sys.stderr)
        return 1

    bids_by_item: dict[int, cs2cap.BatchBidItem] = {
        item.item_id: item for item in bid_response.items
    }
    spread_rows: list[list[str]] = []
    for item_id in item_ids:
        asks = grouped[item_id]
        best_ask = min(asks, key=lambda row: as_int(row.get("lowest_ask")) or 10**12)
        ask = as_int(best_ask.get("lowest_ask"))
        bid_item = bids_by_item.get(item_id)
        best_bid = (
            max(bid_item.quotes, key=lambda quote: quote.highest_bid)
            if bid_item and bid_item.quotes
            else None
        )
        bid = best_bid.highest_bid if best_bid else None
        spread_pct = ((ask - bid) / ask) * 100 if ask and bid else None
        spread_rows.append(
            [
                str(item_id),
                names[item_id],
                as_str(best_ask.get("provider")) or "N/A",
                format_money(ask),
                best_bid.provider if best_bid else "N/A",
                format_money(bid),
                format_percent(spread_pct),
                str(sum(as_int(row.get("quantity")) or 0 for row in asks)),
            ]
        )

    print("Bulk price snapshot + batch bid spread report:")
    print(
        render_table(
            ["item_id", "item", "ask_provider", "ask", "bid_provider", "bid", "spread", "ask_qty"],
            spread_rows[: args.top],
        )
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/examples/pro-tier/create-comparison-table/create_comparison_table.ts
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

import type { BatchBidItem, BatchBidsRequest } from "cs2cap";
import { BidsApi, PricesApi } from "cs2cap";

import { formatApiError, parseNdjsonLines } from "../../_shared/api.js";
import { buildConfiguration, loadApiKey, requireRealApiKey } from "../../_shared/auth.js";
import { parseIntegerOption, printHelpIfRequested } from "../../_shared/cli.js";
import { formatMoney, formatPercent, renderTable } from "../../_shared/render.js";

const SCRIPT_PATH = fileURLToPath(import.meta.url);

function readOptions(): { maxRows: number; top: number } {
  const { values } = parseArgs({
    options: {
      "max-rows": { type: "string" },
      top: { type: "string" },
    },
  });

  return {
    maxRows: parseIntegerOption(values["max-rows"], 500, "--max-rows"),
    top: parseIntegerOption(values.top, 8, "--top"),
  };
}

function numberValue(row: Record<string, unknown>, key: string): number | undefined {
  const value = row[key];
  return typeof value === "number" ? value : undefined;
}

function stringValue(row: Record<string, unknown>, key: string): string | undefined {
  const value = row[key];
  return typeof value === "string" ? value : undefined;
}

async function main(): Promise<number> {
  if (printHelpIfRequested("Usage: npm run example -- examples/pro-tier/create-comparison-table/create_comparison_table.ts [--max-rows 500] [--top 8]")) {
    return 0;
  }
  const args = readOptions();

  let bearerToken: string;
  try {
    bearerToken = loadApiKey("CS2C_API_KEY", SCRIPT_PATH);
    requireRealApiKey(bearerToken);
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    return 1;
  }

  const configuration = buildConfiguration(bearerToken);

  try {
    const pricesApi = new PricesApi(configuration);
    const bidsApi = new BidsApi(configuration);

    const snapshot = await pricesApi.streamFullPricesSnapshot();
    const priceRows = parseNdjsonLines(snapshot, Math.max(args.maxRows, 1));
    const grouped = new Map<number, Array<Record<string, unknown>>>();
    const names = new Map<number, string>();

    for (const row of priceRows) {
      const itemId = numberValue(row, "item_id");
      const name = stringValue(row, "market_hash_name");
      if (itemId == null || name == null) {
        continue;
      }
      grouped.set(itemId, [...(grouped.get(itemId) ?? []), row]);
      names.set(itemId, name);
    }

    const itemIds = Array.from(grouped.keys()).slice(0, Math.max(args.top * 2, args.top));
    const bidResponse = await bidsApi.batchBidLookup({
      batchBidsRequest: { itemIds, currency: "USD" } satisfies BatchBidsRequest,
    });
    const bidsByItem = new Map<number, BatchBidItem>(
      bidResponse.items.map((item) => [item.itemId, item]),
    );

    const rows = itemIds.slice(0, args.top).map((itemId) => {
      const asks = grouped.get(itemId) ?? [];
      const bestAsk = asks.reduce<Record<string, unknown> | undefined>((best, row) => {
        const ask = numberValue(row, "lowest_ask") ?? Number.MAX_SAFE_INTEGER;
        const bestAskValue = best ? numberValue(best, "lowest_ask") ?? Number.MAX_SAFE_INTEGER : Number.MAX_SAFE_INTEGER;
        return ask < bestAskValue ? row : best;
      }, undefined);
      const ask = bestAsk ? numberValue(bestAsk, "lowest_ask") : undefined;
      const bidItem = bidsByItem.get(itemId);
      const bestBid = bidItem?.quotes.reduce((best, quote) =>
        quote.highestBid > best.highestBid ? quote : best,
      );
      const bid = bestBid?.highestBid;
      const spreadPct = ask && bid ? ((ask - bid) / ask) * 100 : undefined;
      return [
        itemId,
        names.get(itemId) ?? "N/A",
        bestAsk ? stringValue(bestAsk, "provider") ?? "N/A" : "N/A",
        formatMoney(ask),
        bestBid?.provider ?? "N/A",
        formatMoney(bid),
        formatPercent(spreadPct),
        asks.reduce((total, row) => total + (numberValue(row, "quantity") ?? 0), 0),
      ];
    });

    console.log("Bulk price snapshot + batch bid spread report:");
    console.log(
      renderTable(
        ["item_id", "item", "ask_provider", "ask", "bid_provider", "bid", "spread", "ask_qty"],
        rows,
      ),
    );
    return 0;
  } catch (error) {
    console.error(await formatApiError(error));
    return 1;
  }
}

void main().then((code) => {
  process.exitCode = code;
});
